File: server/app/utils/TSSecondaryAudioExtractor.py
from biim.mpeg2ts import ts
from biim.mpeg2ts.packetize import packetize_section
from biim.mpeg2ts.parser import SectionParser
from biim.mpeg2ts.pat import PATSection
from biim.mpeg2ts.pmt import PMTSection
from biim.mpeg2ts.section import Section


class TSSecondaryAudioExtractor:
    """MPEG-TS セグメントから第2 AAC 音声と時刻基準だけを抽出する。"""

    AAC_STREAM_TYPES = {0x0F, 0x11}

    @classmethod
    def extract(cls, segment_data: bytes) -> bytes:
        """
        MPEG-TS セグメントから第2 AAC 音声を抽出する。

        Args:
            segment_data (bytes): 映像と複数音声を含む MPEG-TS セグメント

        Returns:
            bytes: PAT、再構築した PMT、第2 AAC、PCR 搬送パケットだけを含む MPEG-TS セグメント

        Raises:
            ValueError: PAT、PMT、または第2 AAC ストリームを取得できない場合
        """

        if len(segment_data) % ts.PACKET_SIZE != 0:
            raise ValueError('MPEG-TS segment size is not aligned to 188-byte packets.')

        # biim のパケットパーサーが受け取れる188バイト単位へ分割し、壊れた同期バイトを先に検出する
        packets = [
            memoryview(segment_data)[offset:offset + ts.PACKET_SIZE]
            for offset in range(0, len(segment_data), ts.PACKET_SIZE)
        ]
        if any(packet[0] != ts.SYNC_BYTE[0] for packet in packets):
            raise ValueError('Invalid MPEG-TS sync byte was found in the segment.')

        # PAT と PMT は biim の SectionParser へ渡し、複数パケットに分割されたセクションも同じ経路で解析する
        pmt_pid = cls.__getPMTPID(packets)
        pmt_section = cls.__getPMTSection(packets, pmt_pid)
        secondary_audio_pid, filtered_pmt_section = cls.__buildFilteredPMT(pmt_section)
        pcr_pid = pmt_section.PCR_PID
        pmt_packet = next((
            packet for packet in packets
            if ts.pid(packet) == pmt_pid and ts.payload_unit_start_indicator(packet)
        ), None)
        if pmt_packet is None:
            raise ValueError('PMT was not found in the MPEG-TS segment.')
        filtered_pmt_packets = packetize_section(
            Section(filtered_pmt_section),
            ts.transport_error_indicator(pmt_packet),
            ts.transport_priority(pmt_packet),
            pmt_pid,
            ts.transport_scrambling_control(pmt_packet),
            ts.continuity_counter(pmt_packet),
        )

        # PAT と再構築した PMT は元の出現位置を保ち、音声 PES の到着順と PCR の時間順を変えずに返す
        # PCR のない入力はそのまま扱い、音声の再生時刻は元の PES に保持する
        output_packets: list[bytes] = []
        for packet in packets:
            pid = ts.pid(packet)
            if pid == 0:
                output_packets.append(bytes(packet))
            elif pid == pmt_pid:
                if packet is pmt_packet:
                    output_packets.extend(filtered_pmt_packets)
            elif pid == secondary_audio_pid:
                output_packets.append(bytes(packet))
            elif pid == pcr_pid and ts.has_pcr(packet):
                output_packets.append(cls.__buildPCROnlyPacket(packet))
        return b''.join(output_packets)

    @staticmethod
    def __getPMTPID(packets: list[memoryview]) -> int:
        """
        PAT から最初の番組の PMT PID を取得する。

        Args:
            packets (list[memoryview]): MPEG-TS パケットのリスト

        Returns:
            int: PMT PID

        Raises:
            ValueError: 有効な PAT または番組の PMT PID を取得できない場合
        """

        pat_parser = SectionParser(PATSection)
        for packet in packets:
            if ts.pid(packet) != 0:
                continue
            pat_parser.push(packet)
            for pat in pat_parser:
                if pat.table_id() != 0x00 or pat.CRC32() != 0:
                    continue
                for program_number, pmt_pid in pat:
                    if program_number != 0:
                        return pmt_pid
        raise ValueError('PAT does not contain a program map PID.')

    @staticmethod
    def __getPMTSection(packets: list[memoryview], pmt_pid: int) -> PMTSection:
        """
        PMT PID のパケットから有効な PMT セクションを取得する。

        Args:
            packets (list[memoryview]): MPEG-TS パケットのリスト
            pmt_pid (int): PAT から取得した PMT PID

        Returns:
            PMTSection: CRC が正しい PMT セクション

        Raises:
            ValueError: 有効な PMT を取得できない場合
        """

        pmt_parser = SectionParser(PMTSection)
        for packet in packets:
            if ts.pid(packet) != pmt_pid:
                continue
            pmt_parser.push(packet)
            for pmt in pmt_parser:
                if pmt.table_id() == 0x02 and pmt.CRC32() == 0:
                    return pmt
        raise ValueError('PMT was not found in the MPEG-TS segment.')

    @classmethod
    def __buildFilteredPMT(cls, pmt: PMTSection) -> tuple[int, bytes]:
        """
        biim が解析した PMT から第2 AAC だけを持つ PMT を構築する。

        Args:
            pmt (PMTSection): 元の PMT セクション

        Returns:
            tuple[int, bytes]: 第2 AAC PID、再構築した PMT セクション

        Raises:
            ValueError: PMT に第2 AAC ストリームがない場合
        """

        aac_streams = [stream for stream in pmt if stream[0] in cls.AAC_STREAM_TYPES]
        if len(aac_streams) < 2:
            raise ValueError('The second AAC stream was not found in the PMT.')

        # 解析済みの記述子長で元セクションをたどり、副音声の ES エントリーをそのまま保持する
        program_info_length = ((pmt[10] & 0x0F) << 8) | pmt[11]
        entry_offset = 12 + program_info_length
        secondary_audio_pid = aac_streams[1][1]
        section_without_crc = bytearray(pmt[:entry_offset])
        for _, pid, descriptors in pmt:
            entry_length = 5 + sum(2 + len(payload) for _, payload in descriptors)
            if pid == secondary_audio_pid:
                section_without_crc.extend(pmt[entry_offset:entry_offset + entry_length])
                break
            entry_offset += entry_length

        # ES を除去した分だけセクション長と CRC を更新する
        section_length = len(section_without_crc) + 4 - 3
        section_without_crc[1] = (section_without_crc[1] & 0xF0) | ((section_length >> 8) & 0x0F)
        section_without_crc[2] = section_length & 0xFF
        crc = Section(section_without_crc).CRC32()
        return secondary_audio_pid, bytes(section_without_crc) + crc.to_bytes(4, 'big')

    @staticmethod
    def __buildPCROnlyPacket(packet: bytes | bytearray | memoryview) -> bytes:
        """
        入力パケットの PCR だけを保持したアダプテーションフィールド専用パケットを返す。

        Args:
            packet (bytes | bytearray | memoryview): PCR を含む MPEG-TS パケット

        Returns:
            bytes: PCR 以外をスタッフィングした MPEG-TS パケット
        """

        # PCR の6バイトをそのまま保持し、映像ペイロードをスタッフィングへ置き換える
        # adaptation field は PCR と不連続フラグだけを持つ構造にそろえる
        output = bytearray(ts.STUFFING_BYTE * ts.PACKET_SIZE)
        output[:12] = packet[:12]
        output[1] &= 0xBF
        output[3] = 0x20 | (packet[3] & 0x0F)
        output[4] = ts.PACKET_SIZE - ts.HEADER_SIZE - 1
        output[5] = 0x10 | (packet[5] & 0x80)
        return bytes(output)

File: client/src/utils/TSSecondaryAudioExtractor.ts
import { calculateCRC32 } from 'arib-mmt-tlv-ts/crc-32.js';
import { readSection } from 'arib-mmt-tlv-ts/ts/si.js';
import { concatBuffers } from 'arib-mmt-tlv-ts/utils.js';

import type { ProgramMapSection } from 'arib-mmt-tlv-ts/ts/si.js';

/** MPEG-TS セグメントから第2 AAC 音声と時刻基準だけを抽出する。 */
export default class MPEGTSSecondaryAudioExtractor {

    private static readonly PACKET_SIZE = 188;
    private static readonly AAC_STREAM_TYPES = new Set([0x0f, 0x11]);

    /**
     * PAT と PMT を解析し、第2 AAC と必要な PCR 搬送パケットだけを返す。
     * @param segmentData 映像と複数音声を含む MPEG-TS セグメント
     * @returns PAT、再構築した PMT、第2 AAC、PCR 搬送パケットだけを含む MPEG-TS セグメント
     */
    static extract(segmentData: Uint8Array): Uint8Array {
        if (segmentData.byteLength % this.PACKET_SIZE !== 0) {
            throw new Error('MPEG-TS segment size is not aligned to 188-byte packets.');
        }

        // 元セグメントをコピーせず、188バイト単位のビューに分割する
        const packets: Uint8Array[] = [];
        for (let offset = 0; offset < segmentData.byteLength; offset += this.PACKET_SIZE) {
            packets.push(segmentData.subarray(offset, offset + this.PACKET_SIZE));
        }
        if (packets.some(packet => packet[0] !== 0x47)) {
            throw new Error('Invalid MPEG-TS sync byte was found in the segment.');
        }

        // セクションの内容と CRC は arib-mmt-tlv-ts で解析し、記述子の元バイト列も保持する
        const {pmtPID, pmt, rawPMT} = this.parseProgramMap(packets);
        const secondaryStream = pmt.streams.filter(stream => this.AAC_STREAM_TYPES.has(stream.streamType))[1];
        if (secondaryStream === undefined) throw new Error('The second AAC stream was not found in the PMT.');

        const pmtPacketIndex = packets.findIndex(packet => this.getPID(packet) === pmtPID && (packet[1] & 0x40) !== 0);
        if (pmtPacketIndex === -1) throw new Error('PMT was not found in the MPEG-TS segment.');
        const filteredPMTSection = this.buildFilteredPMT(rawPMT, secondaryStream.elementaryPID);
        const filteredPMTPackets = this.packetizePMT(
            filteredPMTSection,
            pmtPID,
            packets[pmtPacketIndex],
        );

        // 元のパケット順を維持し、音声の到着順と PCR の時間順が変わらないストリームを組み立てる
        // PCR のない入力はそのまま扱い、音声の再生時刻は元の PES に保持する
        const outputPackets: Uint8Array[] = [];
        packets.forEach((packet, index) => {
            const pid = this.getPID(packet);
            if (pid === 0) outputPackets.push(packet);
            else if (pid === pmtPID && index === pmtPacketIndex) outputPackets.push(...filteredPMTPackets);
            else if (pid === secondaryStream.elementaryPID) {
                outputPackets.push(packet);
            }
            else if (pid === pmt.pcrPID && (packet[3] & 0x20) !== 0 && packet[4] >= 7 && (packet[5] & 0x10) !== 0) {
                outputPackets.push(this.buildPCROnlyPacket(packet));
            }
        });
        return concatBuffers(outputPackets);
    }

    /**
     * TS ヘッダーから PID を取得する。
     * @param packet MPEG-TS パケット
     * @returns PID
     */
    private static getPID(packet: Uint8Array): number {
        return ((packet[1] & 0x1f) << 8) | packet[2];
    }

    /**
     * 指定 PID のセクションを元バイト列のまま取り出す。
     * @param packets MPEG-TS パケット
     * @param pid 対象 PID
     * @returns 結合済み PSI セクション
     */
    private static *readSections(packets: Uint8Array[], pid: number): Generator<Uint8Array> {
        // TSReader の公開イベントには元セクションが含まれないため、書き出しに必要なバイト列だけをここで結合する
        // 内容の解析と CRC 検証は呼び出し側の readSection() に任せる
        let pending: Uint8Array = new Uint8Array();
        let lastCounter: number | null = null;
        for (const packet of packets) {
            if (this.getPID(packet) !== pid) continue;
            // 壊れたパケットや暗号化されたパケットをまたぐセクションは破棄する
            if ((packet[1] & 0x80) !== 0 || (packet[3] & 0xc0) !== 0) {
                pending = new Uint8Array();
                lastCounter = null;
                continue;
            }
            if ((packet[3] & 0x10) === 0) continue;
            const counter = packet[3] & 0x0f;
            if (counter === lastCounter) continue;
            if (lastCounter !== null && counter !== ((lastCounter + 1) & 0x0f)) pending = new Uint8Array();
            lastCounter = counter;
            let offset = 4 + ((packet[3] & 0x20) !== 0 ? 1 + packet[4] : 0);
            if (offset >= packet.length) continue;

            // PUSI パケットではポインター手前が前セクションの末尾、後ろが新セクションの先頭となる
            if ((packet[1] & 0x40) !== 0) {
                const pointer = packet[offset++];
                if (offset + pointer > packet.length) {
                    pending = new Uint8Array();
                    continue;
                }
                if (pending.length > 0) yield concatBuffers([pending, packet.subarray(offset, offset + pointer)]);
                pending = packet.subarray(offset + pointer);
            } else {
                if (pending.length === 0) continue;
                pending = concatBuffers([pending, packet.subarray(offset)]);
            }
            // セクション長までを切り出し、同じパケット内の後続セクションも順に返す
            while (pending.length >= 3 && pending[0] !== 0xff) {
                const length = 3 + (((pending[1] & 0x0f) << 8) | pending[2]);
                if (pending.length < length) break;
                yield pending.subarray(0, length);
                pending = pending.subarray(length);
            }
            if (pending[0] === 0xff) pending = new Uint8Array();
        }
    }

    /**
     * arib-mmt-tlv-ts で PAT と PMT を CRC 検証付きで解析する。
     * @param packets MPEG-TS パケット
     * @returns PAT が参照する PMT PID、解析済み PMT、元セクション
     */
    private static parseProgramMap(packets: Uint8Array[]): {pmtPID: number; pmt: ProgramMapSection; rawPMT: Uint8Array} {
        let pmtPID: number | null = null;
        for (const raw of this.readSections(packets, 0)) {
            const pat = readSection(raw);
            if (pat?.tableId !== 'PAT') continue;
            const program = pat.programs.find(entry => entry.type === 'program');
            if (program?.type === 'program') {
                pmtPID = program.programMapPID;
                break;
            }
        }
        if (pmtPID === null) throw new Error('PAT does not contain a program map PID.');
        for (const rawPMT of this.readSections(packets, pmtPID)) {
            const pmt = readSection(rawPMT);
            if (pmt?.tableId === 'PMT') return {pmtPID, pmt, rawPMT};
        }
        throw new Error('PMT was not found in the MPEG-TS segment.');
    }

    /**
     * 解析済み PMT の元バイト列から第2 AAC のエントリーを残し、CRC32 を更新する。
     * @param rawPMT 元の PMT セクション
     * @param secondaryPID 第2 AAC の PID
     * @returns CRC を含む PMT セクション
     */
    private static buildFilteredPMT(rawPMT: Uint8Array, secondaryPID: number): Uint8Array {
        // ヘッダーと番組記述子、および副音声の ES エントリーを元のままコピーする
        const headerLength = 12 + (((rawPMT[10] & 0x0f) << 8) | rawPMT[11]);
        let entryOffset = headerLength;
        while ((((rawPMT[entryOffset + 1] & 0x1f) << 8) | rawPMT[entryOffset + 2]) !== secondaryPID) {
            entryOffset += 5 + (((rawPMT[entryOffset + 3] & 0x0f) << 8) | rawPMT[entryOffset + 4]);
        }
        const entryLength = 5 + (((rawPMT[entryOffset + 3] & 0x0f) << 8) | rawPMT[entryOffset + 4]);
        const section = concatBuffers([rawPMT.subarray(0, headerLength), rawPMT.subarray(entryOffset, entryOffset + entryLength)]);

        // ES を除去した分だけセクション長と CRC を更新する
        const sectionLength = section.byteLength - 3 + 4;
        section[1] = (section[1] & 0xf0) | ((sectionLength >> 8) & 0x0f);
        section[2] = sectionLength & 0xff;
        const crc = new Uint8Array(4);
        new DataView(crc.buffer).setUint32(0, calculateCRC32(section, -1));
        return concatBuffers([section, crc]);
    }

    /**
     * PMT セクションを必要な数の TS パケットへ格納する。
     * @param section CRC を含む PMT セクション
     * @param pmtPID PMT PID
     * @param sourcePacket 元の PMT パケットヘッダー
     * @returns 再構築した PMT パケットの配列
     */
    private static packetizePMT(
        section: Uint8Array,
        pmtPID: number,
        sourcePacket: Uint8Array,
    ): Uint8Array[] {
        const packets: Uint8Array[] = [];
        let sectionOffset = 0;
        let continuityCounter = sourcePacket[3] & 0x0f;

        // 最初のパケットだけポインターフィールドを置き、後続パケットは184バイトすべてをセクションへ使う
        while (sectionOffset < section.byteLength) {
            const isFirstPacket = sectionOffset === 0;
            const payloadCapacity = this.PACKET_SIZE - 4 - (isFirstPacket ? 1 : 0);
            const sectionChunk = section.subarray(sectionOffset, sectionOffset + payloadCapacity);
            // ライブラリは読み取り専用なので、4バイトの TS ヘッダーとペイロードを直接書き出す
            const packet = new Uint8Array(this.PACKET_SIZE).fill(0xff);
            packet[0] = 0x47;
            packet[1] = (isFirstPacket ? 0x40 : 0) | (sourcePacket[1] & 0x20) | (pmtPID >> 8);
            packet[2] = pmtPID & 0xff;
            packet[3] = (sourcePacket[3] & 0xc0) | 0x10 | continuityCounter;
            if (isFirstPacket) packet[4] = 0;
            packet.set(sectionChunk, isFirstPacket ? 5 : 4);
            packets.push(packet);
            sectionOffset += sectionChunk.byteLength;
            continuityCounter = (continuityCounter + 1) & 0x0f;
        }
        return packets;
    }

    /**
     * 入力パケットの PCR を保持し、映像ペイロードをスタッフィングへ置き換える。
     * @param packet PCR を含む MPEG-TS パケット
     * @returns PCR と不連続フラグを保持した adaptation-field-only パケット
     */
    private static buildPCROnlyPacket(packet: Uint8Array): Uint8Array {
        // PCR の6バイトをそのままコピーし、数値への変換を挟まず元の時刻を保持する
        const output = new Uint8Array(this.PACKET_SIZE).fill(0xff);
        output.set(packet.subarray(0, 12));
        output[1] &= 0xbf;
        output[3] = 0x20 | (packet[3] & 0x0f);
        output[4] = this.PACKET_SIZE - 5;
        output[5] = 0x10 | (packet[5] & 0x80);
        return output;
    }
}
